gcloud-java-storage/src/test/java/com/google/gcloud/storage/BlobWriterChannelTest.java

```diff
@@ -0,0 +1,166 @@
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud.storage;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.gcloud.RetryParams;
+import com.google.gcloud.spi.StorageRpc;
+
+import org.easymock.Capture;
+import org.easymock.EasyMock;
+import org.junit.Test;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Random;
+
+public class BlobWriterChannelTest {
+
+  private static final String BUCKET_NAME = "b";
+  private static final String BLOB_NAME = "n";
+  private static final String UPLOAD_ID = "uploadid";
+  private static final BlobInfo BLOB_INFO = BlobInfo.of(BUCKET_NAME, BLOB_NAME);
+  private static final Map<StorageRpc.Option, ?> EMPTY_RPC_OPTIONS = ImmutableMap.of();
+  private static final int MIN_CHUNK_SIZE = 256 * 1024;
+  private static final int DEFAULT_CHUNK_SIZE = 8 * MIN_CHUNK_SIZE;
+  private static final int CUSTOM_CHUNK_SIZE = 4 * MIN_CHUNK_SIZE;
+  private static final Random RANDOM = new Random();
+
+  private StorageOptions optionsMock;
+  private StorageRpc storageRpcMock;
+  private BlobWriterChannelImpl writer;
+
+  @Before
+  public void setUp() throws IOException, InterruptedException {
+    optionsMock = EasyMock.createMock(StorageOptions.class);
+    storageRpcMock = EasyMock.createMock(StorageRpc.class);
+  }
+
+  @Test
+  public void testCreate() {
+    EasyMock.expect(optionsMock.storageRpc()).andReturn(storageRpcMock);
+    EasyMock.replay(optionsMock);
+    EasyMock.expect(storageRpcMock.open(BLOB_INFO.toPb(), EMPTY_RPC_OPTIONS)).andReturn(UPLOAD_ID);
+    EasyMock.replay(storageRpcMock);
+    writer = new BlobWriterChannelImpl(optionsMock, BLOB_INFO, EMPTY_RPC_OPTIONS);
+    assertTrue(writer.isOpen());
+  }
+
+  @Test
+  public void testWriteWithoutFlush() throws IOException {
+    EasyMock.expect(optionsMock.storageRpc()).andReturn(storageRpcMock);
+    EasyMock.replay(optionsMock);
+    EasyMock.expect(storageRpcMock.open(BLOB_INFO.toPb(), EMPTY_RPC_OPTIONS)).andReturn(UPLOAD_ID);
+    EasyMock.replay(storageRpcMock);
+    writer = new BlobWriterChannelImpl(optionsMock, BLOB_INFO, EMPTY_RPC_OPTIONS);
+    assertEquals(MIN_CHUNK_SIZE, writer.write(ByteBuffer.allocate(MIN_CHUNK_SIZE)));
+  }
+
+  @Test
+  public void testWriteWithFlush() throws IOException {
```

Comment:
I think another useful test-case would be to show how small writes are getting automatically flushed at some point.